lib/mongo/bulk_write.rb

```diff
@@ -154,11 +155,39 @@ def single_statement?(operation)
       SINGLE_STATEMENT_OPS.include?(operation.keys.first)
     end
 
+    # Get the write concern for the bulk write.
+    #
+    # If in a transaction and the bulk write inherits an 
+    # unacknowledged write concern from the collection, remove
+    # the write concern's :w option. Otherwise, return the
+    # unmodified write concern.
+    #
+    # @return [ Mongo::WriteConcern ] The write concern.
+    def applied_write_concern(session)
+      if options[:write_concern]
+        WriteConcern.get(options[:write_concern])
+      else 
+        applied_collection_write_concern(session)
+      end
+    end
+
+    def applied_collection_write_concern(session)
+      wc = collection.write_concern
+      if session.respond_to?(:in_transaction?) && session.in_transaction?
```

Comment:
The pattern we use when session may not be provided is `if session && session.in_transaction?`